README.md
# Clojure-Mutabilidade-com-tomos-e-refs

Codigo desenvolvido em apredizado com base curso Alura: Clojure: Mutabilidade com átomos e refs XD


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-11-17
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-11-17
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hospital/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hospital/compare/0.1.0...0.1.1


src/hospital/aula.clj
(ns hospital.aula
  (:use [clojure pprint])
  (:require [hospital.model :as h.model]
             [hospital.logic :as h.logic]))

(defn simula-um-dia []
  "root binding"
  (def hospital (hospital.model/novo-hospital))
  (def hospital (h.logic/chega-em hospital :espera "111"))
  (def hospital (h.logic/chega-em hospital :espera "222"))
  (def hospital (h.logic/chega-em hospital :espera "333"))
  (clojure.pprint/pprint hospital)



(def hospital (h.logic/chega-em hospital :laboratorio1 "444"))
  (def hospital (h.logic/chega-em hospital :laboratorio3"555"))
  (clojure.pprint/pprint hospital)

  (def hospital (h.logic/atende hospital :laboratorio1))
  (def hospital (h.logic/atende hospital :espera))
  (clojure.pprint/pprint hospital)

  (def hospital (h.logic/chega-em hospital :espera "666"))
  (def hospital (h.logic/chega-em hospital :espera "777"))
  (def hospital (h.logic/chega-em hospital :espera "888"))
  (pprint hospital)
  (def hospital (h.logic/chega-em hospital :espera "999"))
  (clojure.pprint/pprint hospital))

;(pprint hospital)

;(simula-um-dia)
(defn chega-em-malvadao
  [pessoa]
  (def hospital (h.logic/chega-em-pausado hospital :espera pessoa))
  (println "Apos iniciar" pessoa))


;; Muitom claro problema Therad.
(defn simula-um-dia-em-paralelo
  []
  (def hospital (h.model/novo-hospital))
  (.start(Thread. (fn [] (chega-em-malvadao "111"))))
  (.start(Thread. (fn [] (chega-em-malvadao "222"))))
  (.start(Thread. (fn [] (chega-em-malvadao "333"))))
  (.start(Thread. (fn [] (chega-em-malvadao "444"))))
  (.start(Thread. (fn [] (chega-em-malvadao "555"))))
  (.start(Thread. (fn [] (chega-em-malvadao "666"))))
  (.start(Thread. (fn [] (Thread/sleep 4000)
                          (pprint hospital)))))



(simula-um-dia-em-paralelo)




























src/hospital/aula3.clj
(ns hospital.aula3
  (:use [clojure.pprint])
  (:require [hospital.logic :as h.logic]
            [hospital.model :as h.model]))


(defn test-atomao []
  (let [hospital-silveira (atom {:espera h.model/fila_vazia})]
    (println hospital-silveira)
    (pprint hospital-silveira)
    (println @hospital-silveira)

   ; não é assim que altera um conteudo dentro de um átomo!
    (pprint (assoc @hospital-silveira :laboratorio1 h.model/fila_vazia))
    (pprint @hospital-silveira)

    ; Uma das maneiras de alteradas dados de um atomo.
    (swap! hospital-silveira assoc :laboratorio1 h.model/fila_vazia)
    (pprint hospital-silveira)

    (swap! hospital-silveira assoc :laboratorio2 h.model/fila_vazia)
    (pprint hospital-silveira)

    (swap! hospital-silveira update :laboratorio1 conj "111")
    (pprint hospital-silveira)))




;(test-atomao)


(defn chega-em-malvadao!
  [hospital pessoa]
  (swap! hospital h.logic/chega-em-pausado-logando hospital :espera pessoa)
  (println "Apos iniciar" pessoa))



(defn simula-um-dia-em-paralelo
  []
  (let [hospital (atom (h.model/novo-hospital))]
    (.start (Thread. (fn [] (chega-em-malvadao! hospital "111"))))
    (.start (Thread. (fn [] (chega-em-malvadao! hospital"222"))))
    (.start (Thread. (fn [] (chega-em-malvadao! hospital"333"))))
    (.start (Thread. (fn [] (chega-em-malvadao! hospital"444"))))
    (.start (Thread. (fn [] (chega-em-malvadao! hospital"555"))))
    (.start (Thread. (fn [] (chega-em-malvadao! hospital"666"))))
    (.start (Thread. (fn [] (Thread/sleep 8000)
                       (pprint hospital))))))

(simula-um-dia-em-paralelo)

(range 100000)

src/hospital/aula6.clj
(ns hospital.aula6
  (:use [clojure pprint])
  (:require [hospital.model :as h.model]))


(defn cabe-na-fila? [fila]
  (-> fila
      count
      (< 5)))

(defn chega-em
  [fila pessoa]
  (if (cabe-na-fila? fila)
    (conj fila pessoa)
    (throw (ex-info "Fila já está cheia" {:tentando-adicionar pessoa}))))

(defn chega-em!
  "troca de referencia via ref-set"
  [hospital pessoa]
  (let [fila (get hospital :espera)]
    (ref-set fila (chega-em @fila pessoa))))

(defn chega-em!
  "troca de referencia via alter"
  [hospital pessoa]
  (let [fila (get hospital :espera)]
    (alter fila chega-em pessoa)))

(defn simula-um-dia []
  (let [hospital {:espera       (ref h.model/fila-vazia)
                  :laboratorio1 (ref h.model/fila-vazia)
                  :laboratorio2 (ref h.model/fila-vazia)
                  :laboratorio3 (ref h.model/fila-vazia)}]
    (dosync
      (chega-em! hospital "guilherme")
      (chega-em! hospital "maria")
      (chega-em! hospital "lucia")
      (chega-em! hospital "daniela")
      (chega-em! hospital "ana")
      ;(chega-em! hospital "paulo")
      )
    (pprint hospital)))

;(simula-um-dia)


(defn async-chega-em! [hospital pessoa]
  (future
    (Thread/sleep (rand 5000))
    (dosync
      (println "Tentando o codigo sincronizado" pessoa)
      (chega-em! hospital pessoa))))


(defn simula-um-dia-async []
  (let [hospital {:espera       (ref h.model/fila-vazia)
                  :laboratorio1 (ref h.model/fila-vazia)
                  :laboratorio2 (ref h.model/fila-vazia)
                  :laboratorio3 (ref h.model/fila-vazia)}]
    ; simbolo global para você brincar
    (def futures (mapv #(async-chega-em! hospital %) (range 10)))
    (future
      (dotimes [n 4]
        (Thread/sleep 2000)
        (pprint hospital)
        (pprint futures)))))

(simula-um-dia-async)

;(println (future 15))
;(println (future ((Thread/sleep 1000) 15)))


src/hospital/logic.clj
(ns hospital.logic)

(defn cabe-na-fila? [hospital departamento]
  (-> hospital
      (get  ,,, departamento)
      count,,,
      (<,,, 5)))

(defn chega-em
  [hospital departamento pessoa]
  (let [fila (get hospital departamento)
        tamanho-atual (count fila)
        cabe? (< tamanho-atual 5)]
    (if cabe?
      (update hospital departamento conj pessoa)
      (throw (ex-info "Fila já está cheia" { :tentando-adicionar pessoa})))))

(defn chega-em-pausado
  [hospital departamento pessoa]
  (Thread/sleep (*(rand)2000))
  (if (cabe-na-fila? hospital departamento)
    (do (Thread/sleep 1000))
      (update hospital departamento conj pessoa))
      (throw (ex-info "Fila já está cheia" { :tentando-adicionar pessoa})))


(defn chega-em-pausado-logando
  [hospital departamento pessoa]
  (println "tentando adcionar esta pessoa"pessoa)
  (Thread/sleep (* (rand)2000))
  (if (cabe-na-fila? hospital departamento)
    (do
      ;(Thread/sleep 1000))
    (update hospital departamento conj pessoa)
      (println "Dei update" pessoa))
  (throw (ex-info "Fila já está cheia" { :tentando-adicionar pessoa}))))


(defn atende
  [hospital departamento]
  (update hospital departamento pop))

(defn proxima
  [hospital departamento]
  (-> hospital
      departamento
      peek))

(defn transfere
  [hospital de para]
  (let [pessoa (peek (get hospital de))]
    (-> hospital
        (atende de)
        (chega-em para pessoa))))

(defn atende-completo
  "somente para demosntrar que é possivel retornar os dois (quem e a fila restante.)"
  [hospital departamento]
  {:paciente (update hospital departamento peek)
   :hospital (update hospital departamento pop)})

(defn atende-completo-que-chama-ambos
  [hospital departamento]
  (let [fila (get hospital departamento)
        peek-pop (juxt peek pop)
        [pessoa fila-atualizada] (peek-pop fila)
        hospital-atualizado (update assoc departamento fila-atualizada)]
    {:paciente pessoa
     :hospital hospital-atualizado}))

src/hospital/aula4.clj
(ns hospital.aula4
  (:use [clojure pprint])
  (:require [hospital.logic :as h.logic]
            [hospital.model :as h.model]))

(defn chega-sem-malvado! [hospital pessoa]
  (swap! hospital h.logic/chega-em :espera pessoa)
  (println "apos inserir" pessoa))

#_(defn simula-um-dia-em-paralelo-com-mapv
  "simulação utilizando mapv para forçar quase que imperativamente a execução do que era lazy"
    []
  (let [hospital (atom (h.model/novo-hospital))
        pessoas ["111" "222" "333" "444" "555" "666"]]
    (mapv #(.start (Thread. (fn [] (chega-sem-malvado! hospital %))))pessoas)

                       (pprint hospital)))


;_________________________________________________________________________________________________
(defn simula-um-dia-em-paralelo-com-mapv-refatorada
  []
  (let [hospital (atom (h.model/novo-hospital))
        pessoas ["111" "222" "333" "444" "555" "666"]
        starta-thread-de-chegada #(.start (Thread. (fn [] (chega-sem-malvado! hospital %))))]

    (mapv starta-thread-de-chegada pessoas)

          (.start (Thread. (fn [] (Thread/sleep 8000)
                             (pprint hospital))))))

(simula-um-dia-em-paralelo-com-mapv-refatorada)

;___________________________________________________________________________________
(defn starta-thread-de-chegada
  [hospital pessoa]
  (.start (Thread. (fn [] (chega-sem-malvado! hospital pessoa)))))


(defn simula-um-dia-em-paralelo-com-mapv-com-partial
  []
  (let [hospital (atom (h.model/novo-hospital))
        pessoas ["111" "222" "333" "444" "555" "666"]
        starta (partial starta-thread-de-chegada pessoas)]

    (mapv starta pessoas)

    (.start (Thread. (fn [] (Thread/sleep 8000)
                       (pprint hospital))))))

(simula-um-dia-em-paralelo-com-mapv-com-partial)
;______________________________________________________________________________________________________
(defn simula-um-dia-em-paralelo-com-doseq
  "executa a sequencia"
  []
  (let [hospital (atom (h.model/novo-hospital))
        pessoas (range 6)]
    (doseq [pessoas pessoas]
      (starta-thread-de-chegada hospital pessoas))

    (.start(Thread.(fn [] (Thread/sleep 8000)
                     (pprint hospital))))))

(simula-um-dia-em-paralelo-com-doseq)
;________________________________________________________________________________________________________
(defn simula-um-dia-em-paralelo-com-dotimes
  "Realmente estou preocupado em executar N vezes"
  []
  (let [hospital (atom (h.model/novo-hospital))]
    (
      dotimes [pessoas 6]
      (starta-thread-de-chegada hospital pessoas))

    (.start(Thread.(fn [] (Thread/sleep 8000)
                     (pprint hospital))))))

(simula-um-dia-em-paralelo-com-dotimes)